src/sybild/sybild.proto

```diff
@@ -0,0 +1,185 @@
+
+// Sybil is an OLAP analytics store.
+service Sybil {
+  // Ingest inserts new data into a sybil dataset.
+  rpc Ingest (IngestRequest) returns (IngestResponse) {}
+  // Query retrieves results from sybil.
+  rpc Query (QueryRequest) returns (QueryResponse) {}
+  // ListTables lists the tables in a sybil database.
+  rpc ListTables (ListTablesRequest) returns (ListTablesResponse) {}
+  // GetTable gets details about a particular table.
+  rpc GetTable (GetTableRequest) returns (Table) {}
+
+  // todo //rpc Expire
+  // todo //rpc Trim
+}
+
+// IngestRequest inserts records into a sybil dataset.
+message IngestRequest {
+  // dataset is the name of the dataset.
+  string dataset = 1;
+
+  // records is the set of records to insert.
+  repeated google.protobuf.Struct records = 2;
+}
+
+// IngestResponse reports the results of an ingestion request.
+message IngestResponse {
+  int64 number_inserted = 1;
+}
+
+// QueryType defines the types of query that can be performed.
+enum QueryType {
+  QUERY_TYPE_UNKNOWN = 0;
+  TABLE = 1;
+  TIME_SERIES = 2;
+  DISTRIBUTION = 3;
+  SAMPLES = 4;
+}
+
+// QueryOp
+enum QueryOp {
+  QUERY_OP_UNKNOWN = 0;
+  AVERAGE = 1;
+  HISTOGRAM = 2;
+}
+
+// QueryFilterOp
+enum QueryFilterOp {
+  QUERY_FILTER_OP_UNKNOWN = 0;
+  RE = 1;
+  NRE = 2;
+  EQ = 3;
+  NEQ = 4;
+  GT = 5;
+  LT = 6;
+  IN = 7;
+  NIN = 8;
+}
+
+// QueryFilter
+message QueryFilter {
+  string column = 1;
+  QueryFilterOp op = 2;
+  string value = 3;
+}
+
+// QueryRequest
+message QueryRequest {
+  // dataset is the name of the dataset.
+  string dataset = 1;
+  QueryType type = 2;
+  int64 limit = 3;
+  repeated string ints = 4;
+  repeated string strs = 5;
+  repeated string groups = 6;
+  repeated string distinct_by = 7;
+  string sort = 8;
+  // todo: prune_(by,sort)
+
+  string time_column = 9;
+  int64 time_bucket = 10;
+
+  QueryOp op = 11;
+  bool log_histogram = 12;
+
+  repeated QueryFilter int_filters = 13;
+  repeated QueryFilter str_filters = 14;
+  repeated QueryFilter set_filters = 15;
+
+  int64 int_histogram_bucket_size = 16;
+  // todo: query cache control
+  // todo: ingestion log reading option
+}
+
+// Histogram
+message Histogram {
+  double mean = 1;
+  int64 max = 2;
+  int64 min = 3;
+  int64 total_count = 4;
+
+  repeated int64 percentiles = 5;
+  map<int64, int64> int_buckets = 6;
+
+  int64 range_start = 7;
+  int64 range_end = 8;
+  double std_deviation = 9;
+}
+
+// QueryResult
+message QueryResult {
+  map<string, Histogram> histograms = 1;
+  // todo //Distinct *hll.LogLogBeta = 2;
+
+  string GroupByKey = 3;
+  int64 Count = 4;
+  int64 Samples = 5;
+}
+
+// ResultMap
+message ResultMap {
+  map<string, QueryResult> values = 1;
+}
+
+// SetField
+message SetField {
+  repeated string values = 1;
+}
+
+// Record
+message Record {
+  map<string, string> strs = 1;
+  map<string, int64> ints = 2;
+  map<string, SetField> sets = 3;
+
+  // todo //string path = 4;
```

Comment:
@okayzed can you think of why this would be useful to surface? I couldn't offhand but would appreciate your input.

Reply:
no, it's likely a remnant. it's similar to timestamp field in that it gets filled out when loading the record off disk, but it was used for session queries